server/src/test/java/org/elasticsearch/cluster/metadata/IndexMetadataTests.java

```diff
@@ -698,6 +604,179 @@ public void testReshardingBWCSerialization() throws IOException {
         assertEquals(idx, IndexMetadata.builder(deserialized).reshardingMetadata(reshardingMetadata).build());
     }
 
+    public void testIndexShardCountToXContent() throws IOException {
+        int expectedShards = randomIntBetween(0, 100);
+        IndexMetadata.IndexShardCount count = new IndexMetadata.IndexShardCount(expectedShards);
+        XContentBuilder builder = XContentFactory.jsonBuilder();
+        builder.startObject();
+        count.toXContent(builder, null);
+        builder.endObject();
+        assertTrue(Strings.toString(builder).contains("\"shard_count\":" + expectedShards));
+    }
+
+    public void testIndexShardCountBuilder() {
+        int expectedShards = randomIntBetween(0, 100);
+        IndexMetadata.IndexShardCount count = IndexMetadata.IndexShardCount.builder().setCount(expectedShards).build();
+        assertEquals(expectedShards, count.getCount());
+    }
+
+    public void testFromIndexMetaDataWithValidIndexMetaDataObject() throws IOException {
+        int numberOfShards = randomFrom(1, 2, 4, 8, 16);
+        IndexMetadata indexMetadata = randomIndexMetadata(numberOfShards);
+
+        final XContentBuilder builder = JsonXContent.contentBuilder();
+        builder.startObject();
+        IndexMetadata.FORMAT.toXContent(builder, indexMetadata);
+        builder.endObject();
+        XContentParser parser = createParser(builder);
+
+        IndexMetadata.IndexShardCount count = IndexMetadata.IndexShardCount.fromIndexMetaData(parser);
+        assertEquals(numberOfShards, count.getCount());
+    }
+
+    public void testFromIndexMetaDataWithValidIndexMetaDataObjectWithoutEventIngestedField() throws IOException {
+        int numberOfShards = randomFrom(1, 2, 4, 8, 16);
+        IndexMetadata indexMetadata = randomIndexMetadata(numberOfShards);
+
+        final XContentBuilder builder = JsonXContent.contentBuilder();
+        builder.startObject();
+        IndexMetadata.FORMAT.toXContent(builder, indexMetadata);
+        builder.endObject();
+        XContentParser parser = createParser(builder);
+
+        // convert XContent to a map and remove the IndexMetadata.KEY_EVENT_INGESTED_RANGE entry
+        // to simulate IndexMetadata from an older cluster version (before TransportVersions.EVENT_INGESTED_RANGE_IN_CLUSTER_STATE)
+        Map<String, Object> indexMetadataMap = XContentHelper.convertToMap(BytesReference.bytes(builder), true, XContentType.JSON).v2();
+        removeEventIngestedField(indexMetadataMap);
+
+        IndexMetadata.IndexShardCount count = IndexMetadata.IndexShardCount.fromIndexMetaData(parser);
+        assertEquals(numberOfShards, count.getCount());
+    }
+
+    public void testFromIndexMetaDataWithoutNumberOfShardsSettingReturnsNegativeOne() throws IOException {
```

Comment:
For these legacy versions, I'd rather we had an actual copy of the index metadata blob written by the older version and verified that we read it as expected. Otherwise if we make a change to `IndexMetadata` serialization and a parallel change to the `IndexShardCount` deserialization then we may not notice a problem with deserializing actual blobs until it's too late. I mean it's ok to synthesize the older versions like this too, but using a real copy of some ancient metadata would be a valuable addition.

Reply:
Ah that's a good point, we could add something to `org.elasticsearch.upgrades.MultiVersionRepositoryAccessIT` to check that we can load the shard count from the blob actually written by the old cluster. We almost cover this already since e.g. `testUpgradeMovesRepoToNewMetaVersion` uses the current version to delete snapshots created by a much older version, except that I think we delete _all_ the snapshots for the old index in one go so we never actually look at its shard count.

However those tests only go back as far as 8.0.0 AFAICT. That's actually a testing gap, I think we need to be able to deal with snapshot repositories going back as far as 5.0.0 (via the `old-lucene-versions` plugin for pre-8.0.0 indices).